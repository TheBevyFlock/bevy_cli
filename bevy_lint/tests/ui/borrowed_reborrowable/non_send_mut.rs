//! This tests the `borrowed_reborrowable` lint, specifically when triggered on the `NonSendMut` type.

#![feature(register_tool)]
#![register_tool(bevy)]
#![deny(bevy::borrowed_reborrowable)]

use bevy::prelude::*;

#[derive(Resource)]
struct MyResource;

// OK: Lint does not apply to immutable references
fn immutable_reference(_param: &NonSendMut<MyResource>) {
    // ...
}

//~| HELP: use `NonSendMut` instead
//~v ERROR: parameter takes `&mut NonSendMut` instead of a re-borrowed `NonSendMut`
fn mutable_reference(_param: &mut NonSendMut<MyResource>) {
    // ...
}

//~| HELP: use `NonSendMut` instead
//~v ERROR: parameter takes `&mut NonSendMut` instead of a re-borrowed `NonSendMut`
fn mutable_reference_return<'a>(_param: &'a mut NonSendMut<MyResource>) -> usize {
    123
}

// OK: Lint does not apply when return type relies on reference lifetime
fn mutable_reference_bounded_return<'a>(param: &'a mut NonSendMut<MyResource>) -> &'a MyResource {
    &*param
}

// OK: Lint does not apply when return type relies on reference lifetime
fn mutable_reference_bounded_return_complex<'a>(
    param: &'a mut NonSendMut<MyResource>,
) -> Vec<(usize, &'a MyResource)> {
    vec![(1, &*param)]
}

fn main() {
    fn some_system(mut param: NonSendMut<MyResource>) {
        immutable_reference(&param);
        mutable_reference(&mut param);
        _ = mutable_reference_return(&mut param);
        _ = mutable_reference_bounded_return(&mut param);
        _ = mutable_reference_bounded_return_complex(&mut param);
    }

    App::new().add_systems(Update, some_system).run();
}
