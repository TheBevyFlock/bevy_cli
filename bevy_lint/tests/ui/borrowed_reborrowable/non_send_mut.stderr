error: parameter takes `&mut NonSendMut` instead of a re-borrowed `NonSendMut`
  --> tests/ui/borrowed_reborrowable/non_send_mut.rs:19:22
   |
19 | fn mutable_reference(_param: &mut NonSendMut<MyResource>) {
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use `NonSendMut` instead: `mut _param: bevy::prelude::NonSendMut<'_, MyResource>`
   |
note: the lint level is defined here
  --> tests/ui/borrowed_reborrowable/non_send_mut.rs:5:9
   |
5  | #![deny(bevy::borrowed_reborrowable)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: parameter takes `&mut NonSendMut` instead of a re-borrowed `NonSendMut`
  --> tests/ui/borrowed_reborrowable/non_send_mut.rs:25:33
   |
25 | fn mutable_reference_return<'a>(_param: &'a mut NonSendMut<MyResource>) -> usize {
   |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use `NonSendMut` instead: `mut _param: bevy::prelude::NonSendMut<'_, MyResource>`

error: aborting due to 2 previous errors

